tests/url-validation.py

```diff
@@ -0,0 +1,57 @@
+import os
+import json
+import requests
+from pprint import pprint
+import threading
+
+repository_directory = os.getcwd()
+domains_directory = os.path.join(repository_directory, "domains")
+threads = []
+
+def has_url_field(file_path):
+    with open(file_path, "r") as file:
+        data = json.load(file)
+        record = data.get("record")
+        if record and "URL" in record:
+            return record["URL"]
+    return None
+
+def is_url_reachable(url: str):
+    try:
+        response = requests.head(url, allow_redirects=True)
+        return response.status_code // 100 in [1, 2, 3]  # Check if status code is in the 1xx or 2xx or 3xx range (success)
+    except requests.exceptions.RequestException:
+        return False
+
+urls_data = {
+    "valid": {},
+    "invalid": {},
+    "non-http": {}
+}
+
+def handle_url_validation(root, filename):
+    file_path = os.path.join(root, filename)
+    url: str = has_url_field(file_path)
+    if url:
+        if url.startswith("http://") or url.startswith("https://"):
+            if is_url_reachable(url):
+                urls_data["valid"][file_path] = url
+                print(f"URL '{url}' in file '{file_path}' is reachable.")
+            else:
+                urls_data["invalid"][file_path] = url
+                print(f"URL '{url}' in file '{file_path}' is not reachable.")
+        else:
+            urls_data["non-http"][file_path] = url
+            print(f"URL '{url}' in file '{file_path}' is neither HTTP nor HTTPS.")
+
+
+for root, _, files in os.walk(domains_directory):
+    for filename in files:
+        thread = threading.Thread(target=handle_url_validation, args=(root, filename))
```

Comment:
Creating 1000s of physical threads is a terrible idea even for a one-shot script. Can we instead use some sort of batching that processes things maybe 10-20 at a time?

Reply:
#8145